src/shell-utils.ts

```diff
@@ -53,27 +52,71 @@ export async function createAndRunShell(
+        readline.emitKeypressEvents(process.stdin);
+        if (process.stdin.isTTY) {
+            process.stdin.setRawMode(true);
+        }
+
+        // Force stdin to be in flowing mode in case the stream was paused.
+        // The stream is paused when connect() is called from quickstart due
+        // to the way the prompt library works.
+        if (process.stdin.readableFlowing === false) {
+            process.stdin.resume();
+        }
+
+        // Max input delay to wait, in ms
+        let maxInputDelay = 1;
+        let previousInput = Date.now();
+        const maxInputDelayLimit = 101;
+
+        // To keep track of this is a large stdin buffer (i.e. copy paste)
+        let inputBuffer: string[] = [];
+        let bufferFunction: NodeJS.Timeout = null;
 
-            // Force stdin to be in flowing mode in case the stream was paused.
-            // The stream is paused when connect() is called from quickstart due
-            // to the way the prompt library works.
-            if (process.stdin.readableFlowing === false) {
-                process.stdin.resume();
+        process.stdin.on('keypress', async (_, key) => {
+            // Implement some custom logic for batching input
+            // Ref: https://stackoverflow.com/questions/66755705/detect-pasted-input-with-readline-nodejs
+
+            // Add our input to our array of input
+            inputBuffer.push(key.sequence);
+
+            // Keep increasing maxInputDelay if the last input was less than 5ms ago
+            // We cap this wait at maxInputDelayLimitms
+            if ((Date.now() - previousInput) < 5) {
+                // Only increase if our maxInputDelay < maxInputDelayLimit
+                if (maxInputDelay < maxInputDelayLimit) {
+                    maxInputDelay += 5;
+                }
+            } else {
+                // Else reset our delay
+                maxInputDelay = 1;
             }
 
-            process.stdin.on('keypress', (_, key) => { observer.next(key.sequence); });
-        });
-        source.pipe(bufferTime(50), filter(buffer => buffer.length > 0)).subscribe(keypresses => {
-            // This pipe is in order to allow copy-pasted input to be treated like a single string
-            terminal.writeString(keypresses.join(''));
+            // Update when we got our last input
+            previousInput = Date.now();
+
+            // If we get a new input, clear the timeout function
+            if (bufferFunction === null) {
+                // send the input to a function after a certain amount of time has passed
+                bufferFunction = setTimeout(() => {
+                    // Loop over the array, and send it as chunks of 10000
+                    // Otherwise we get keysplitting/general errors if we try to send too much data
+                    const chunk = 10000;
+                    for (let i = 0,j = inputBuffer.length; i < j; i += chunk) {
```

Comment:
small simplification
```suggestion
                    for (let i = 0; i < inputBuffer.length; i += chunk) {
```

Reply:
https://github.com/bastionzero/zli/pull/242/commits/e29980c2b38ddd3c49cdbd8639857fa2aed36d55